src/Domain/CoordinateMaps/KerrHorizon.cpp

```diff
@@ -0,0 +1,177 @@
+  coords_dot_spin = dot(source_coords, spin_);
+  subexpr_1 = 4. * source_coords_sq * (1. - fac);
+  subexpr_2 = 2. * fac * coords_dot_spin;
+
+  for (size_t i = 0; i < 3; ++i) {
+    gsl::at(dfac_dx, i) = subexpr_1 * gsl::at(source_coords, i) +
+                          2. * spin_mag_sq_ * gsl::at(source_coords, i) -
+                          subexpr_2 * gsl::at(spin_, i);
+  }
+  dfac_dx = dfac_dx / (square(source_coords_sq) + square(coords_dot_spin));
+
+  const ReturnType sqrt_fac = sqrt(fac);
+
+  // not mathematically a part of `dfac_dx` but can be absorbed to avoid
+  // allocation for temporary
+  dfac_dx = dfac_dx / (2. * sqrt_fac);
+
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j < 3; ++j) {
+      jac.get(i, j) = gsl::at(dfac_dx, j) * gsl::at(source_coords, i);
+    }
+    jac.get(i, i) += sqrt_fac;
+  }
+  return jac;
+}
+
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>
+KerrHorizon::inv_jacobian(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+
+  tnsr::Ij<ReturnType, 3, Frame::NoFrame> inv_jac(
+      get_size(dereference_wrapper(source_coords[0])));
+
+  ReturnType& mapped_mag_sq = get<0, 0>(inv_jac);
+  ReturnType& mapped_mag = get<0, 1>(inv_jac);
+  ReturnType& mapped_dot_spin = get<0, 2>(inv_jac);
+  ReturnType& mapped_sq_min_spin_sq = get<1, 0>(inv_jac);
+  ReturnType& r = get<1, 1>(inv_jac);
+  ReturnType& fac = get<1, 2>(inv_jac);
+  std::array<ReturnType, 3> dr_dx{};
+  if constexpr (std::is_same_v<ReturnType, DataVector>) {
+    dr_dx[0].set_data_ref(&get<2, 0>(inv_jac));
+    dr_dx[1].set_data_ref(&get<2, 1>(inv_jac));
+    dr_dx[2].set_data_ref(&get<2, 2>(inv_jac));
+  }
+
+  auto mapped = operator()(source_coords);
+
+  mapped_mag_sq = dot(mapped, mapped);
+  mapped_mag = sqrt(mapped_mag_sq);
+  mapped_dot_spin = dot(mapped, spin_);
+  mapped_sq_min_spin_sq = mapped_mag_sq - spin_mag_sq_;
+  r = sqrt(0.5 * (mapped_sq_min_spin_sq + sqrt(square(mapped_sq_min_spin_sq) +
+                                               4 * square(mapped_dot_spin))));
+  fac = 1. / (2. * r * r * r - mapped_sq_min_spin_sq * r);
+
+  for (size_t i = 0; i < 3; ++i) {
+    gsl::at(dr_dx, i) =
+        (r * r * mapped.at(i) + mapped_dot_spin * gsl::at(spin_, i)) * fac;
+  }
+
+  // normalized from this point
+  mapped = mapped / mapped_mag;
+  const ReturnType r_by_mapped = r / mapped_mag;
+
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j < 3; ++j) {
+      inv_jac.get(i, j) = gsl::at(dr_dx, j) * gsl::at(mapped, i) -
+                          r_by_mapped * gsl::at(mapped, i) * gsl::at(mapped, j);
+    }
+    inv_jac.get(i, i) += r_by_mapped;
+  }
+  return inv_jac;
+}
+
+template <typename T>
+tt::remove_cvref_wrap_t<T> KerrHorizon::stretch_factor(
+    const std::array<T, 3>& source_coords) const noexcept {
+  const auto coords_mag_sq = dot(source_coords, source_coords);
```

Comment:
are you sure you want this to be an expression template? It seems like the ideal in terms of both memory allocations and not recomputing things would be to pass in a `gsl::not_null<T*>` that gets populated, though I'd be fine with just making this a `T` for now and adding a comment along the lines of `// Note: this is doing a memory allocation that could be elided if necessary, but since this map isn't time-dependent it's unclear how important that is without profiling`

Reply:
I have removed the allocation but at a large readability cost for `operator()`

Reply:
Yea, the unfortunate thing I've learned is that the really ugly C-style (ish) code is actually often much faster than the nicer looking/easier to read code :(